File: engine/lab/cloze_lab/scheduler/blocks.py
"""Block manager and masks (DESIGN.md §5.4): semi-autoregressive block diffusion.

Generate the output left->right in blocks of length L. Within the active block,
diffusion runs over [prompt + frozen blocks ‖ active block]; finalized blocks
freeze. The attention is **block-causal** — the *one-way law* (DESIGN): a
position attends to its own block and every earlier one, never forward. That is
what keeps frozen-block K/V exact across later blocks (the basis for the Tier B
cache, §5.5): a frozen block's attention pattern — and so its K/V — is identical
whether or not later blocks exist.

``block_len = 0`` is whole-sequence mode: a single block, fully bidirectional
(prompt and output mutually attend), best for infill, no inter-block structure.
This module is pure logic; numpy is used only to build the boolean mask.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from cloze_lab.models.base import BoolArray


@dataclass(frozen=True, slots=True)
class Block:
    """One output block: board positions [start, end)."""

    index: int
    start: int
    end: int

    @property
    def span(self) -> tuple[int, int]:
        return (self.start, self.end)


@dataclass(frozen=True, slots=True)
class BlockPlan:
    """Left-to-right blocks over the output region [prompt_len, prompt_len+max_new).

    ``block_len = 0`` => one whole-sequence block. ``block_len > 0`` => semi-AR
    blocks of that length (the last may be shorter).
    """

    prompt_len: int
    max_new: int
    block_len: int

    def __post_init__(self) -> None:
        if self.prompt_len < 1:
            raise ValueError(f"prompt_len must be >= 1, got {self.prompt_len}")
        if self.max_new < 1:
            raise ValueError(f"max_new must be >= 1, got {self.max_new}")
        if self.block_len < 0:
            raise ValueError(f"block_len must be >= 0 (0 = whole-sequence), got {self.block_len}")

    @property
    def whole_sequence(self) -> bool:
        return self.block_len == 0

    def blocks(self) -> list[Block]:
        start, end = self.prompt_len, self.prompt_len + self.max_new
        if self.block_len == 0:
            return [Block(index=0, start=start, end=end)]
        out: list[Block] = []
        pos = start
        while pos < end:
            out.append(Block(index=len(out), start=pos, end=min(pos + self.block_len, end)))
            pos += self.block_len
        return out


def block_id(pos: int, prompt_len: int, block_len: int) -> int:
    """-1 for prompt positions; 0, 1, 2, ... for successive output blocks."""
    if pos < prompt_len:
        return -1
    return (pos - prompt_len) // block_len


def attention_mask(working_len: int, prompt_len: int, block_len: int) -> BoolArray:
    """[working_len, working_len] bool mask; ``M[q, k]`` True means q may attend to k.

    whole-sequence (block_len=0): fully bidirectional (all True). Block mode:
    ``M[q, k] = block_id(k) <= block_id(q)`` — the prompt attends only to itself,
    each block attends to the prompt and all earlier blocks plus bidirectionally
    within itself, and nothing attends forward (the one-way law).
    """
    if block_len == 0:
        return np.ones((working_len, working_len), dtype=bool)
    ids = np.array(
        [block_id(p, prompt_len, block_len) for p in range(working_len)], dtype=np.int64
    )
    return ids[None, :] <= ids[:, None]  # M[q, k] = ids[k] <= ids[q]

File: engine/core/src/blocks.cpp
// blocks.cpp — implementation of cloze/blocks.hpp, mirroring blocks.py.
#include "cloze/blocks.hpp"

#include <algorithm>
#include <stdexcept>

namespace cloze {

std::vector<Block> BlockPlan::blocks() const {
    if (prompt_len < 1) throw std::invalid_argument("prompt_len must be >= 1");
    if (max_new < 1) throw std::invalid_argument("max_new must be >= 1");
    if (block_len < 0) throw std::invalid_argument("block_len must be >= 0 (0 = whole-sequence)");

    const int start = prompt_len;
    const int end = prompt_len + max_new;
    std::vector<Block> out;
    if (block_len == 0) {
        out.push_back(Block{0, start, end});
        return out;
    }
    for (int pos = start; pos < end; pos += block_len) {
        const int b_end = (pos + block_len < end) ? pos + block_len : end;
        out.push_back(Block{static_cast<int>(out.size()), pos, b_end});
    }
    return out;
}

int block_id(int pos, int prompt_len, int block_len) {
    if (pos < prompt_len) return -1;
    return (pos - prompt_len) / block_len;
}

Mask attention_mask(int working_len, int prompt_len, int block_len) {
    Mask m;
    m.n = working_len;
    m.data.assign(static_cast<size_t>(working_len) * working_len, 0);
    if (block_len == 0) {
        // fully bidirectional: all attend.
        std::fill(m.data.begin(), m.data.end(), static_cast<unsigned char>(1));
        return m;
    }
    std::vector<int> ids(working_len);
    for (int p = 0; p < working_len; ++p) ids[p] = block_id(p, prompt_len, block_len);
    // M[q, k] = ids[k] <= ids[q]  (the one-way law).
    for (int q = 0; q < working_len; ++q)
        for (int k = 0; k < working_len; ++k)
            m.data[static_cast<size_t>(q) * working_len + k] = (ids[k] <= ids[q]) ? 1 : 0;
    return m;
}

}  // namespace cloze
